tests/interop/tls-shim/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+#![allow(async_fn_in_trait)]
+
+use std::{error::Error, fmt::Debug};
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+use common::{InteropTest, CLIENT_GREETING, LARGE_DATA_DOWNLOAD_GB, SERVER_RESPONSE};
+
+pub mod s2n_tls_shim;
+
+const ONE_MB: usize = 1_000_000;
+const ONE_GB: usize = 1_000_000_000;
+
+/// The ServerTLS trait allows for shared code between s2n-tls, rustls,
+/// and openssl. All of these TLS implementations have relatively similar API shapes
+/// which this trait attempts to abstract over.
+pub trait ServerTLS<T> {
+    type Config;
+    type Acceptor: Clone + Send + 'static;
+    // the Stream is generic to allow for Turmoil test usage
+    type Stream: Send + AsyncRead + AsyncWrite + Debug + Unpin;
+
+    fn get_server_config(test: InteropTest) -> Result<Option<Self::Config>, Box<dyn Error>>;
+
+    fn acceptor(config: Self::Config) -> Self::Acceptor;
+
+    fn accept(
+        server: &Self::Acceptor,
+        transport_stream: T,
+    ) -> impl std::future::Future<Output = Result<Self::Stream, Box<dyn Error + Send + Sync>>> + Send;
+
+    /// `handle_server_connection` provides generic "handle connection" functionality.
+    /// It will automatically implement correct application behavior for tests that
+    /// don't require any implementation specific apis.
+    async fn handle_server_connection(
+        test: InteropTest,
+        mut stream: Self::Stream,
+    ) -> Result<(), Box<dyn Error + Send + Sync>> {
+        tracing::info!("Executing the {:?} scenario", test);
+        match test {
+            InteropTest::Handshake => {
+                // no application data exchange in the handshake case
+            }
+            InteropTest::RequestResponse | InteropTest::MTLSRequestResponse => {
+                let mut client_greeting_buffer = vec![0; CLIENT_GREETING.as_bytes().len()];
+                stream.read_exact(&mut client_greeting_buffer).await?;
+                assert_eq!(client_greeting_buffer, CLIENT_GREETING.as_bytes());
+
+                stream.write_all(SERVER_RESPONSE.as_bytes()).await?;
+            }
+            InteropTest::LargeDataDownload => {
+                let mut client_greeting_buffer = vec![0; CLIENT_GREETING.as_bytes().len()];
+                stream.read_exact(&mut client_greeting_buffer).await?;
+                assert_eq!(client_greeting_buffer, CLIENT_GREETING.as_bytes());
+
+                let mut data_buffer = vec![0; ONE_MB];
+                // for each GB
+                for i in 0..LARGE_DATA_DOWNLOAD_GB {
+                    if i % 10 == 0 {
+                        tracing::info!("GB sent: {}", i);
+                    }
+                    data_buffer[0] = (i % u8::MAX as u64) as u8;
+                    for _ in 0..(ONE_GB / ONE_MB) {
+                        stream.write_all(&data_buffer).await?;
+                    }
+                }
+            }
+            InteropTest::LargeDataDownloadWithFrequentKeyUpdates => {
```

Comment:
Trying to think through how this would work with resumption. Where would that logic live 🤔? This test only allows for a single connection currently.

Reply:
I think you'd end up essentially special-casing the logic for the client. If the test is `InteropTest::Resumption` then you would call the `handle_session_resumption` method twice. The implementation would be responsible for 
1. storing the session ticket from the first connection
2. setting the session ticket on the second connection.

The config for session resumption would set the `SessionTicketCallback` and the `ConnectionInitializer`.